sdk/rust/oak/src/io/receiver.rs

```diff
@@ -66,14 +66,29 @@ impl<T: Decodable> Receiver<T> {
     }
 
     /// Wait for a value to be available from this receiver.
+    ///
+    /// Returns [`ChannelReadStatus`] of the wrapped handle, or `Err(OakStatus::ErrTerminated)` if
+    /// the Oak Runtime is terminating,
     #[allow(clippy::trivially_copy_pass_by_ref)]
-    pub fn wait(&self) -> Result<(), OakStatus> {
+    pub fn wait(&self) -> Result<ChannelReadStatus, OakStatus> {
         // TODO(#500): Consider creating the handle notification space once and for all in `new`.
         let read_handles = vec![self.handle];
         let mut space = crate::new_handle_space(&read_handles);
         crate::prep_handle_space(&mut space);
         let status =
             unsafe { oak_abi::wait_on_channels(space.as_mut_ptr(), read_handles.len() as u32) };
-        crate::result_from_status(status as i32, ())
+
+        match crate::result_from_status(status as i32, ()) {
+            Err(OakStatus::ErrTerminated) => Err(OakStatus::ErrTerminated),
+            _ => match space
+                .get(oak_abi::SPACE_BYTES_PER_HANDLE - 1)
+                .cloned()
+                .map(i32::from)
+                .and_then(ChannelReadStatus::from_i32)
+            {
+                Some(channel_status) => Ok(channel_status),
+                None => unreachable!("wait_on_channels always yields a valid channel read status if the returned status is not Err(OakStatus::ErrTerminated)."),
```

Comment:
Reading the documentation of `unreachable`, it feels like it is meant for much more self-contained impossible conditions (e.g. truly unreachable match arms, code after a loop that never terminates, etc): https://doc.rust-lang.org/std/macro.unreachable.html

I don't think it should be used for this kind of precondition checks, for which panic seems the default choice.

Alternatively we may decide to log at `error` level, since that's reserved for logic errors. I guess it is worth having a quick discussion on panic vs error on a GitHub issue, could you start one? For now, feel free to use what you prefer, it should be easy to later on grep for `error!` or `panic!` and make sure they are consistent, hopefully.